### PR TITLE
Add Ipa appliance cli tests, fix ssh

### PR DESCRIPTION
**Pull request by jaryn**
Purpose or Intent
=================

__Fixing__ ssh client currently deadlocking when some output written to stderr.
__Improving__ ssh client -- removed sleep
__Adding tests__ for configuration logging-in using IPA

### CODE REVIEW

File: cfme/utils/ssh.py [hunk 1]
```diff
@@ -323,58 +324,36 @@ def _run_command(self, command, timeout=RUNCMD_TIMEOUT, reraise=False, ensure_ho
...
+
+            stdin.close()
+            channel.shutdown_write()
+            while not channel.closed:
+                readq, _, _ = select.select([channel], [], [], timeout)
+                for c in readq:
+                    if c.recv_ready():
+                        write_output(channel.recv(len(c.in_buffer)), self.f_stdout)
+                    if c.recv_stderr_ready():
+                        write_output(channel.recv(len(c.in_buffer)), self.f_stderr)
```
Comment (RonnyPfannschmidt) [inline cfme/utils/ssh.py]: **required** the correct apis here are `in_stderr_buffer` and `recv_stderr`

Comment (RonnyPfannschmidt) [reply cfme/utils/ssh.py]: **OPTIONAL** the need for the channel files got removed, i believe its absolutely fine just to go for the channel

File: cfme/tests/cli/test_appliance_cli.py [hunk 2]
```diff
@@ -199,3 +202,35 @@ def is_failover_started(appliance):
     wait_for(is_failover_started, func_args=[apps[2]], timeout=450, handle_exception=True)
     apps[2].wait_for_evm_service()
     apps[2].wait_for_web_ui()
+
+
+def test_appliance_console_cli_configure_ipa(configured_appliance):
+    auth_provider_conf = auth_data.auth_providers.freeipa01
```
Comment (mshriver) [inline cfme/tests/cli/test_appliance_cli.py]: This is referencing a specific key within auth_providers yaml, and is not appropriate - it won't be available to upstream community and needs to be wrapped in a fixture that finds an appropriate server for test.

File: cfme/tests/cli/test_appliance_cli.py [hunk 3]
```diff
@@ -199,3 +202,35 @@ def is_failover_started(appliance):
     wait_for(is_failover_started, func_args=[apps[2]], timeout=450, handle_exception=True)
     apps[2].wait_for_evm_service()
     apps[2].wait_for_web_ui()
+
+
+def test_appliance_console_cli_configure_ipa(configured_appliance):
```
Comment (mshriver) [inline cfme/tests/cli/test_appliance_cli.py]: This test is already covered by the following module, which includes parametrization for the available IPA servers:

https://github.com/ManageIQ/integration_tests/blob/master/cfme/tests/integration/test_cfme_auth.py

File: cfme/tests/cli/test_appliance_cli.py [hunk 4]
```diff
@@ -199,3 +202,35 @@ def is_failover_started(appliance):
...
+    view, _ = wait_for(lambda: navigate_to(configured_appliance.server, 'Authentication'),
+                       handle_exception=True, delay=10, timeout=120)
+    view.form.auth_mode.fill('External (httpd)')
+    view.form.httpd_role.fill("x")
+    view.save.click()
+
+    login_page = navigate_to(configured_appliance.server, 'LoginScreen')
+    ipa_users = [user for user in auth_data.test_data.test_users
+                 if 'freeipa01' in user.providers]
+    user = ipa_users[0]
```
Comment (mshriver) [inline cfme/tests/cli/test_appliance_cli.py]: When testing login as different users, its best (required pretty much) to use the context manager.  Please see existing auth tests (cfme/tests/integration/test_cfme_auth) for use of this context manager with a user object. It sets the appliance user to the given user object, and any login/logout done within the context is done as the given user. At exit of the context, the admin user is restored as the default for the appliance.

Without the context controlling this, a failure within these statements would leave the appliance logged in as the non-admin user, impacting subsequent tests.

**Comment by dajoRH** [comment]
Lint report for commit 7663dee421511720c12219ac4437ef955a2fd4d5:

`cfme/utils/ssh.py`:
- :bangbang: Line 337:28: **E702** *multiple statements on one line (semicolon)*
- :red_circle: Line 351:17: **F841** *local variable 'stderr' is assigned to but never used*
- :red_circle: Line 352:17: **F841** *local variable 'stdout' is assigned to but never used*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/tests/cli/test_appliance_cli.py (hunk 3 above)
Comment (jaryn) [reply cfme/tests/cli/test_appliance_cli.py]: @mshriver I cannot see that the module is configuring the use of IPA using the appliance_cli, which is the main thing my test is testing.

Comment (mshriver) [reply cfme/tests/cli/test_appliance_cli.py]: It is, via this Appliance method, which is in the `ApplianceConsoleCli` class:
https://github.com/ManageIQ/integration_tests/blob/master/cfme/utils/appliance/__init__.py#L183

The IPAAuthProvider class includes a method that supplies a fill dictionary that can be used with the above method:
https://github.com/ManageIQ/integration_tests/blob/master/cfme/utils/auth/__init__.py#L257

**Comment by jaryn** [comment]
It is a dup

**Pull request closed without merging**